quickwit-directories/src/bundle_storage.rs

```diff
@@ -0,0 +1,289 @@
+}
+
+impl BundleStorage {
+    /// Creates a new BundleStorage.
+    ///
+    /// The provided data must include the footer_bytes at the end of the slice, but it can have more.
+    pub fn new(
+        storage: Arc<dyn Storage>,
+        bundle_file_name: PathBuf,
+        meta_data: &[u8],
+    ) -> io::Result<Self> {
+        let metadata = BundleStorageMetadata::open(meta_data)?;
+        Ok(BundleStorage {
+            storage,
+            bundle_file_name,
+            metadata,
+        })
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct BundleStorageMetadata {
+    pub(crate) files: HashMap<PathBuf, FileOffsets>,
+}
+
+impl BundleStorageMetadata {
+    fn open(data: &[u8]) -> io::Result<Self> {
+        let footer_size: u64 = BinarySerializable::deserialize(&mut &data[data.len() - 8..])?;
+
+        let footer_slice = &data[data.len() - 8 - footer_size as usize..data.len() - 8];
+        //println!("{}", String::from_utf8(footer_slice.to_owned()).unwrap());
+        serde_json::from_slice(footer_slice)
+            .map_err(|err| io::Error::new(ErrorKind::InvalidData, err))
+    }
+    fn get(&self, path: &Path) -> StorageResult<FileOffsets> {
+        self.files
+            .get(path)
+            .cloned()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "not found in metadata").into())
+    }
+    fn exists(&self, path: &Path) -> bool {
+        self.files.contains_key(path)
+    }
+}
+
+#[async_trait]
+impl Storage for BundleStorage {
+    async fn put(
+        &self,
+        path: &Path,
+        _payload: quickwit_storage::PutPayload,
+    ) -> quickwit_storage::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn copy_to_file(
+        &self,
+        path: &Path,
+        _output_path: &Path,
+    ) -> quickwit_storage::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(
+        &self,
+        path: &Path,
+        range: Range<usize>,
+    ) -> quickwit_storage::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> quickwit_storage::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> quickwit_storage::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> quickwit_storage::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
```

Comment:
Checking bundle_file_name seems strange as this is the bundle.

I would check only files in metadata.

Reply:
we could check if the bundle exists in the underlying storage, if  the bundle doesn't exist the file does not exist. but it's much more expensive